### PR TITLE
ci(workflows): Add PR package versioning and publish to GitHub Packages

### PR DESCRIPTION
**Pull request by kploch**
### CI: PR package versioning and publishing

- **PR builds** now produce NuGet packages with branch-identifying prerelease tags (e.g. `3.1.1-pr.42.fix-serialization`) and publish them to GitHub Packages
- **Master push builds** continue publishing prerelease packages (`3.1.1-prerelease`) to GitHub Packages only — unchanged
- **Release builds** continue publishing stable packages to NuGet.org — unchanged
- Build and test steps now consistently use Release configuration
- Master publish condition tightened (was `always()`, now gated on push + master)
- GitHub Pages deployment failure no longer blocks package publishing (`continue-on-error: true`)
- Release workflow: creates GitHub Release before the version bump commit for idempotent reruns with deterministic timestamps

#### Version matrix

| Scenario | Example Version | Published To |
|----------|----------------|-------------|
| PR build | `3.1.2-pr.187.fix-serialization` | GitHub Packages |
| Master push | `3.1.2-prerelease` | GitHub Packages |
| Release | `3.1.0` | NuGet.org |

### Fix: IsInExtensions equality check and null guard

- **Bug fix:** `In` and `NotIn` extension methods now correctly match default values (e.g. `0`, `null`) by using `EqualityComparer<TValue?>.Default.Equals` instead of `value.Equals(v)` which skipped default values
- **Null guard:** Added `comparer.NotNull()` validation for `IComparer`-based overloads
- **Refactor:** Changed static method calls to extension method syntax for consistency
- **Test coverage:** Added comprehensive tests for enum values, `IEnumerable` overloads, `IComparer` overloads, and `NotIn` variants

### How PR versioning works

For PR builds, the workflow:
1. Reads the current NBGV base version (e.g. `3.1`)
2. Sanitises the branch name for NuGet SemVer2 compatibility
3. Guards against empty/numeric-only sanitised identifiers
4. Calls `dotnet nbgv set-version` to embed PR number + branch name in the prerelease tag
5. Builds and publishes packages to GitHub Packages

## Issue ticket number and link

N/A — infrastructure improvement and bug fix

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] Do we need to implement analytics? — N/A
- [x] Will this be part of a product update? If yes, please write one phrase about this update.
  - CI: PR builds now publish testable NuGet packages to GitHub Packages
  - Fix: `In`/`NotIn` extension methods now correctly handle default values and null comparers

## No breaking changes

All public API signatures remain unchanged. The `IsInExtensions` equality fix corrects behaviour (default values now match correctly), which is a bug fix rather than a breaking change.

___

## **CodeAnt-AI Description**
**Publish branch-specific PR packages and fix matching for default values**

### What Changed
- PR builds now publish NuGet packages to GitHub Packages with a branch- and PR-specific prerelease label, making preview packages easy to identify and test.
- Build and test runs now use Release settings, and master publishing only runs on master push builds.
- If GitHub Pages deployment fails, package publishing can still continue.
- Release runs now create the GitHub Release before the follow-up version bump, and release commits use a stable timestamp for repeatable reruns.
- `In` and `NotIn` now handle `null`, `0`, and other default values correctly, and comparer-based checks fail fast when no comparer is provided.
- Added test coverage for enum values, list-based checks, comparer-based checks, and `NotIn` cases.

### Impact
`✅ Easier testing of PR packages`
`✅ Fewer release publishing interruptions`
`✅ Correct matching for null and zero values`
<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: .github/workflows/build-dotnet.yml [hunk 1]
```diff
@@ -125,14 +142,29 @@ jobs:
...
+      - name: Publish PR packages to GitHub Packages
+        if: success() && github.event_name == 'pull_request'
         env:
           GH_PACKAGES_TOKEN: ${{ secrets.GH_PACKAGES_TOKEN }}
         run: |
+          if [ -z "$GH_PACKAGES_TOKEN" ]; then
+            echo "::warning::GH_PACKAGES_TOKEN not available, skipping PR package publish"
+            exit 0
+          fi
+          dotnet nuget add source --username kploch --password "$GH_PACKAGES_TOKEN" --store-password-in-clear-text --name github "https://nuget.pkg.github.com/mrploch/index.json"
```
[the remark this answers is not in the document]
Comment (kploch) [reply .github/workflows/build-dotnet.yml:181]: False positive — `--store-password-in-clear-text` is required on Linux runners for `dotnet nuget add source` (it fails without it). The GitHub Actions runner is ephemeral and destroyed immediately after the job completes, so the NuGet config file never persists beyond the single run. The existing master publish step uses the same flag and this is the standard pattern in all dotnet GitHub Actions workflows.

File: .github/workflows/build-dotnet.yml [hunk 2]
```diff
@@ -125,14 +142,29 @@ jobs:
...
         env:
           GH_PACKAGES_TOKEN: ${{ secrets.GH_PACKAGES_TOKEN }}
         run: dotnet nuget add source --username kploch --password "$GH_PACKAGES_TOKEN" --store-password-in-clear-text --name github "https://nuget.pkg.github.com/mrploch/index.json"
       - name: Publish NuGet packages to GitHub Packages
-        if: always()
+        if: success() && github.event_name == 'push' && github.ref == 'refs/heads/master'
+        env:
+          GH_PACKAGES_TOKEN: ${{ secrets.GH_PACKAGES_TOKEN }}
+        run: |
+          dotnet nuget push ./**/*.nupkg --source github --skip-duplicate -k "$GH_PACKAGES_TOKEN"
```
[the remark this answers is not in the document]
Comment (kploch) [reply .github/workflows/build-dotnet.yml:167]: False positive — this project uses `GeneratePackageOnBuild=true` in `Directory.Build.props`, which produces `.nupkg` files inside each project's `bin/Release/` directory (not a centralised `./artifacts` folder). The `./**/*.nupkg` glob is the correct and standard pattern for this build setup. `--skip-duplicate` handles any edge cases, and test projects have `IsPackable=false` so they don't produce `.nupkg` files anyway.

File: .github/workflows/build-dotnet.yml [hunk 3]
```diff
@@ -125,14 +142,29 @@ jobs:
       # Publish prerelease NuGet packages to GitHub Packages
       # Release packages are published to nuget.org via the release workflow
       - name: Add GitHub Packages source
-        if: always()
+        if: success() && github.event_name == 'push' && github.ref == 'refs/heads/master'
```
[the remark this answers is not in the document]
Comment (kploch) [reply .github/workflows/build-dotnet.yml:158]: Valid point — the `Deploy GitHub Pages` step didn't have `continue-on-error: true`, so a transient Pages failure would have blocked package publishing. Fixed by adding `continue-on-error: true` to the Pages deployment step in the latest commit.

File: .github/workflows/build-dotnet.yml [hunk 4]
```diff
@@ -39,6 +39,29 @@ jobs:
...
+          PR_NUMBER: ${{ github.event.pull_request.number }}
+          BRANCH_NAME: ${{ github.head_ref }}
+        run: |
+          # Sanitize branch name for NuGet SemVer2 prerelease identifiers [a-zA-Z0-9-]
+          SANITIZED=$(echo "$BRANCH_NAME" | sed 's/[^a-zA-Z0-9-]/-/g' | sed 's/--*/-/g' | sed 's/^-//' | sed 's/-$//' | cut -c1-20 | sed 's/-$//')
+          # Read current major.minor from NBGV
+          CURRENT_VERSION=$(dotnet nbgv get-version --variable SimpleVersion)
+          BASE=$(echo "$CURRENT_VERSION" | cut -d. -f1-2)
+          # Override the prerelease tag to include PR number and branch name
+          dotnet nbgv set-version "${BASE}-pr.${PR_NUMBER}.${SANITIZED}"
```
[the remark this answers is not in the document]
Comment (kploch) [reply .github/workflows/build-dotnet.yml:63]: Good catch — while unlikely in practice (GitHub branch names always contain non-numeric characters), this is a reasonable defensive guard. Applied the suggested hardening in the latest commit.

File: change-log/2026-03-15-ci-pr-package-versioning.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+## CI: PR package versioning and publishing
+
+- **PR builds** now publish NuGet packages to GitHub Packages with a branch-identifying prerelease tag (e.g. `3.1.1-pr.42.fix-serialization`), making it easy to test packages from in-progress work.
+- **Master push builds** continue to publish prerelease packages (`3.1.1-prerelease`) to GitHub Packages only.
+- **Release builds** continue to publish stable packages to NuGet.org via the manual release workflow.
```
[the remark this answers is not in the document]
Comment (kploch) [reply change-log/2026-03-15-ci-pr-package-versioning.md]: Good suggestion — added a version example for release builds for consistency. Fixed in the latest commit.

File: change-log/2026-03-15-ci-pr-package-versioning.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+## CI: PR package versioning and publishing
+
+- **PR builds** now publish NuGet packages to GitHub Packages with a branch-identifying prerelease tag (e.g. `3.1.1-pr.42.fix-serialization`), making it easy to test packages from in-progress work.
+- **Master push builds** continue to publish prerelease packages (`3.1.1-prerelease`) to GitHub Packages only.
+- **Release builds** continue to publish stable packages (e.g. `3.1.0`) to NuGet.org via the manual release workflow.
```
[the remark this answers is not in the document]
Comment (kploch) [reply change-log/2026-03-15-ci-pr-package-versioning.md:5]: False positive — the version examples are illustrative; the actual patch number depends on the git commit height at any given point, so both are valid examples. The intent is to show the versioning *pattern* (prerelease tag structure), not pin to a specific commit height.

File: src/Common/IsInExtensions.cs [hunk 7]
```diff
@@ -35,7 +35,7 @@ public static bool In<TValue>(this TValue? value, IEnumerable<TValue?> values)
     {
         values.NotNull(nameof(values));
 
-        return values.Any(v => value.IsNotDefault() && value!.Equals(v));
+        return values.Any(v => EqualityComparer<TValue?>.Default.Equals(value, v));
     }
 
     /// <summary>
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/IsInExtensions.cs:41]: False positive — the use of `IComparer<T>` here is intentional. `StringComparer` (the primary use case) implements both `IComparer<string>` and `IEqualityComparer<string>`, and `Compare(a, b) == 0` is a well-known pattern for equality checks via ordering comparers.

Using `IComparer` rather than `IEqualityComparer` was a deliberate design choice: it allows callers to pass `StringComparer.OrdinalIgnoreCase`, `StringComparer.Create(culture, options)`, etc. directly without needing a separate overload set. If we added `IEqualityComparer` overloads, we'd likely need both interface variants anyway, which doubles the API surface for little gain.

The semantic risk (ordering equivalence ≠ strict equality) is acknowledged but acceptable for the "is this value in this set?" use case.

File: src/Common/IsInExtensions.cs [hunk 8]
```diff
@@ -49,7 +49,7 @@ public static bool In<TValue>(this TValue? value, IEnumerable<TValue?> values)
     ///     <c>true</c> if the <paramref name="value" /> is equal to one of the <paramref name="values" /> according to the <paramref name="comparer" />, <c>false</c>
     ///     otherwise.
     /// </returns>
-    public static bool In<TValue>(this TValue? value, IComparer<TValue?> comparer, params TValue[] values) => In(value, comparer, (IEnumerable<TValue>)values);
+    public static bool In<TValue>(this TValue? value, IComparer<TValue?> comparer, params TValue[] values) => value.In(comparer, (IEnumerable<TValue>)values);
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/IsInExtensions.cs:52]: Valid point — added `comparer.NotNull(nameof(comparer))` in the canonical `In(IComparer, IEnumerable)` overload (line 68), which all `params` overloads and `NotIn` variants delegate to. Fixed in c1ca548.

File: src/Common/IsInExtensions.cs [hunk 9]
```diff
@@ -119,5 +119,5 @@ public static bool NotIn<TValue>(this TValue? value, IComparer<TValue?> comparer
     ///     <c>false</c>
     ///     otherwise.
     /// </returns>
-    public static bool NotIn<TValue>(this TValue? value, IComparer<TValue?> comparer, IEnumerable<TValue?> values) => !In(value, comparer, values);
+    public static bool NotIn<TValue>(this TValue? value, IComparer<TValue?> comparer, IEnumerable<TValue?> values) => !value.In(comparer, values);
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/IsInExtensions.cs:123]: Valid concern, but already covered — `NotIn(comparer, IEnumerable)` delegates directly to `In(comparer, values)`, which now has a `comparer.NotNull()` guard (added in c1ca548). Adding a redundant guard in `NotIn` would validate the same parameter twice on every call. The fail-fast behaviour is the same: `ArgumentNullException` with the parameter name.

File: src/Common/IsInExtensions.cs [hunk 10]
```diff
@@ -35,7 +35,7 @@ public static bool In<TValue>(this TValue? value, IEnumerable<TValue?> values)
     {
         values.NotNull(nameof(values));
 
-        return values.Any(v => value.IsNotDefault() && value!.Equals(v));
+        return values.Any(v => EqualityComparer<TValue?>.Default.Equals(value, v));
```
[the remark this answers is not in the document]
Comment (kploch) [reply src/Common/IsInExtensions.cs:38]: False positive — this comment describes the fix that was already applied. The current code uses `EqualityComparer<TValue?>.Default.Equals(value, v)` which correctly handles default values (null, 0, etc.) for all types.

**Comment by kploch** [comment]
Addressed the CodeRabbit and Bito test cleanup suggestions: refactored the four separate enum `[Fact]` test methods into two parameterised `[Theory]` tests with `[InlineData]`, and removed the unused variable assignments and `Console.WriteLine` debug statements. Fixed in dd2f1b9.
